yt/yt/client/table_client/composite_compare.cpp

```diff
@@ -276,4 +308,112 @@ TFingerprint CompositeFarmHash(TYsonStringBuf value)
+        bool isInsideList = unclosedListCount - (item.GetType() == EYsonItemType::EndList);
+        auto resultingItemSize = GetMinResultingSize(item, /*isInsideList*/ isInsideList);
+
+        if (resultingItemSize > remainingBytes) {
+            break;
+        }
+
+        bool canBeTruncatedFurther = true;
+
+        switch (item.GetType()) {
+            case EYsonItemType::BeginList:
+                ++unclosedListCount;
+                writer.WriteBeginList();
+                break;
+            case EYsonItemType::EndList:
+                --unclosedListCount;
+                writer.WriteEndList();
+                break;
+            case EYsonItemType::EntityValue:
+                writer.WriteEntity();
+                break;
+            case EYsonItemType::Int64Value:
+                writer.WriteBinaryInt64(item.UncheckedAsInt64());
+                break;
+            case EYsonItemType::Uint64Value:
+                writer.WriteBinaryUint64(item.UncheckedAsUint64());
+                break;
+            case EYsonItemType::DoubleValue:
+                writer.WriteBinaryDouble(item.UncheckedAsDouble());
+                break;
+            case EYsonItemType::BooleanValue:
+                writer.WriteBinaryBoolean(item.UncheckedAsBoolean());
+                break;
+            case EYsonItemType::StringValue: {
+                auto truncatedString = item.UncheckedAsString().Trunc(remainingBytes - resultingItemSize);
+                writer.WriteBinaryString(truncatedString);
+                // This is a slight overestimation, since a smaller string might have a shorter varint part storing its length.
+                resultingItemSize += truncatedString.size();
+                break;
+            }
+            // Maps and attributes are not comparable.
+            // However, we can store everything up to this point into the truncated string.
+            case EYsonItemType::BeginAttributes:
+            case EYsonItemType::EndAttributes:
+            case EYsonItemType::BeginMap:
+            case EYsonItemType::EndMap:
+            case EYsonItemType::EndOfStream:
+                canBeTruncatedFurther = false;
+        }
+
+        if (!canBeTruncatedFurther) {
+            break;
+        }
+
+        nonEmptyResult = true;
+
+        if (unclosedListCount && item.GetType() != EYsonItemType::BeginList) {
+            writer.WriteItemSeparator();
+        }
+
+        remainingBytes -= resultingItemSize;
+    }
+
+    YT_VERIFY(unclosedListCount >= 0);
+    for (; unclosedListCount;) {
+        writer.WriteEndList();
+        if (--unclosedListCount) {
+            writer.WriteItemSeparator();
+        }
+    }
+
+    if (nonEmptyResult) {
+        writer.Finish();
+    } else {
+        return {};
+    }
+
+    // TODO(achulkov2): Let's double check that we haven't missed any cases. I would also be fine with a slight overflow here.
+    YT_ASSERT(std::ssize(truncatedYson) <= size);
```

Comment:
Why not verify?

Reply:
I am a bit scared of this check in production. Considering that it is not critical that this check holds.

Reply:
Changed to YT_LOG_ALERT.